odl/set/pspace.py

```diff
@@ -458,6 +466,85 @@ def size(self):
         """The number of factors of this vector's space."""
         return self.space.size
 
+    @property
+    def total_size(self):
+        """The total number of elements of this space's factors"""
+        # Todo: only calculate once?
+        total = 0
+        for part in self.parts:
+            if isinstance(part, ProductSpaceVector):
+                total += part.total_size
+            else:
+                total += part.size
+        return total
+
+    @property
+    def shape(self):
+        """The shape of this vector, given by (size, )."""
+        return (self.size,)
+
+    @property
+    def ndim(self):
+        """The number of dimensions of this vector, always 1."""
+        return 1
+
+    def __len__(self):
+        """``v.__len__() <==> len(v)``."""
+        return len(self.space)
+
+    @property
+    def dtype(self):
+        """The dtype of the vector elements.
+
+        Uses `numpy.find_common_type` if no common dtype exists.
+
+        See also
+        --------
+        common_dtype
+        """
+        # Todo: only calculate once?
+        return np.find_common_type(self.parts, [])
+
+    def asarray(self, out=None):
+        """Representation as linear numpy array.
+
+        Returns
+        -------
+        asarray : `numpy.ndarray`
+            Array representation
+        """
+        if out is None:
+            out = np.empty(self.total_size, dtype=self.dtype)
+        else:
+            assert isinstance(out, np.ndarray)
+            assert out.size == self.total_size
+            assert np.can_cast(self.dtype, out.dtype)
+
+        index = 0
+        for part in self.parts:
+            if isinstance(part, ProductSpaceVector):
+                out[index:index + part.total_size] = part.asarray()
+                index += part.total_size
+            else:
+                out[index:index + part.size] = part.asarray()
+                index += part.size
+
```

Comment:
I'm trying to understand how this works. If `part` is some linear vector, then this is all fine. What happens if `part` is a `DiscreteLpVector` on some higher-dimensional grid? Then the statement in 529 would fail. I guess it should be `asflatarray()` there?